VenafiTppPS/Code/Public/Grant-TppToken.ps1

```diff
@@ -0,0 +1,145 @@
+<#
+.SYNOPSIS
+Get an API Access and Refresh Token from TPP for use in New-TPPSession (or other scripts/utilities that take such a token)
+
+.DESCRIPTION
+Accepts username/password credential, scope, and ClientId to get a token grant from specified TPP server.
+
+.PARAMETER ServerUrl
+URL for the Venafi server.
+
+.PARAMETER Credential
+Username / password credential used to request API Token
+
+.PARAMETER ClientId
+Applcation Id configured in Venafi for token-based authentication
+
+.PARAMETER Scope
+Hashtable with Scopes and privilege restrictions.
+The key is the scope and the value is one or more privilege restrictions separated by commas.
+
+.EXAMPLE
+Grant-TPPToken -ServerUrl 'https://mytppserver.example.com' -Scope @{ Certificate = "manage,discover"; Config = "manage" } -ClientId 'MyAppId' -Credential $credential
+
+#>
+function Grant-TppToken {
+
+    [CmdletBinding()]
+
+    param (
```

Comment:
On purpose. I worry about giving people a way to just get a "*.*" scoped token (and they have to match what is entered into Aperture anyway).  Long-term, it would be better to be able to document each function with the scopes that they need, so that we are able to provide meaningful guidance on errors if they're missing a required scope or even be able to produce a report on what scopes a given script uses.

I am not married to this, but felt like if we made it easy to do the wrong thing (like just create applications that had all the scopes then people would take the easy way). A lot of those scopes (like "restricted" should only really be used if one knows what they're doing and why!)

Reply:
I'm not married to it either :)  I typically try and make things easier for folks, but in this case I can see the concern.

As of now, if the proper scope/restriction hasn't been granted for the token being used, the error will tell you the missing scope, but not restriction.  I've added an idea @ https://ideas.venafi.com/forums/916237-platform/suggestions/40579540-add-privilege-restriction-required-to-error-messag.  If we could get the platform to provide this info as opposed to the module having to maintain it, that would be great.